File: python/samples/02_partial_discover.py
#!/usr/bin/env python3
"""Sample 02 — Partial discovery (single service tree walk).

Demonstrates:
  - ClientContext.discover_async(service="Systems")
  - DiscoveryResult.service_uri()
  - Re-running partial discovery for a second service

Usage:
    python 02_partial_discover.py [--host HOST] [--port PORT]
"""

from __future__ import annotations

import argparse
import asyncio
import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parents[1]))

import redfish_sdk
from redfish_sdk import AuthMode, ConnectionConfig, Credentials
from redfish_sdk.errors import RedfishConnectionError


def parse_args() -> argparse.Namespace:
    p = argparse.ArgumentParser(description="Sample 02 — Partial Discovery")
    p.add_argument("--host", default="127.0.0.1")
    p.add_argument("--port", type=int, default=8000)
    p.add_argument("--user", default="admin")
    p.add_argument("--password", default="password")
    p.add_argument("--no-tls-verify", action="store_true")
    return p.parse_args()


async def main() -> None:
    args = parse_args()

    creds = Credentials(username=args.user, password=args.password)
    config = ConnectionConfig(verify_tls=not args.no_tls_verify)

    try:
        ctx = await redfish_sdk.connect_async(
            host=args.host,
            port=args.port,
            credentials=creds,
            auth_mode=AuthMode.SESSION,
            config=config,
        )
    except RedfishConnectionError as exc:
        print(f"[ERROR] {exc}")
        sys.exit(1)

    async with ctx:
        # ── Partial: only walk Systems ──────────────────────────────────
        print("Partial discovery — Systems only …")
        result = await ctx.discover_async(service="Systems")

        systems_uri = result.service_uri("Systems")
        if systems_uri:
            print(f"  Systems URI: {systems_uri}")
        else:
            print("  Systems not found on this endpoint")

        # ── Partial: EventService ───────────────────────────────────────
        print("\nPartial discovery — EventService only …")
        result2 = await ctx.discover_async(service="EventService")

        events_uri = result2.service_uri("EventService")
        if events_uri:
            print(f"  EventService URI: {events_uri}")
        else:
            print("  EventService not available on this endpoint")

        # ── Root info only ──────────────────────────────────────────────
        print("\nFetching root info (no tree walk) …")
        root_result = await ctx.discover_async(root_only=True)
        caps = root_result.capabilities
        print(f"  Redfish version : {caps.redfish_version}")
        print(f"  Product         : {caps.product}")

        print("\n✓ Partial discovery complete")


if __name__ == "__main__":
    asyncio.run(main())

File: rust/samples/src/bin/02_partial_discover.rs
use redfish_sdk::{connect, AuthMode, ConnectionConfig, Credentials, RedfishError};

fn parse_args() -> (String, u16, String, String, bool) {
    let mut host = "127.0.0.1".to_string();
    let mut port = 8000u16;
    let mut user = "admin".to_string();
    let mut password = "password".to_string();
    let mut no_tls = false;
    let mut it = std::env::args().skip(1);
    while let Some(a) = it.next() {
        match a.as_str() {
            "--host"          => { host = it.next().unwrap(); }
            "--port"          => { port = it.next().unwrap().parse().unwrap(); }
            "--user"          => { user = it.next().unwrap(); }
            "--password"      => { password = it.next().unwrap(); }
            "--no-tls-verify" => { no_tls = true; }
            _ => {}
        }
    }
    (host, port, user, password, no_tls)
}

#[tokio::main]
async fn main() -> Result<(), RedfishError> {
    tracing_subscriber::fmt::init();
    let (host, port, user, password, no_tls) = parse_args();
    let no_tls_proto = std::env::args().any(|a| a == "--no-tls");
    let creds = Credentials::new(&user, &password);
    let config = ConnectionConfig { verify_tls: !no_tls, use_tls: !no_tls_proto, ..Default::default() };
    let ctx = connect(&host, port, creds, AuthMode::Session, config).await?;

    println!("Partial discovery - Systems only ...");
    let result = ctx.discovery().partial("Systems").await?;
    match result.service_uri("Systems") {
        Some(uri) => println!("  Systems URI: {uri}"),
        None      => println!("  Systems not found"),
    }

    println!("
Partial discovery - EventService only ...");
    let result2 = ctx.discovery().partial("EventService").await?;
    match result2.service_uri("EventService") {
        Some(uri) => println!("  EventService URI: {uri}"),
        None      => println!("  EventService not available"),
    }

    println!("
Fetching root info ...");
    let _ = ctx.discovery().root().await?;
    let caps = ctx.capabilities();
    println!("  Redfish version : {}", caps.redfish_version);

    println!("
Partial discovery complete");
    Ok(())
}
